File: research/explorations/bench-realistic/guards.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

//go:noinline
func classify(x int64) int64 {
	if x >= 900 { return 30 }
	if x >= 700 { return 25 }
	if x >= 500 { return 20 }
	if x >= 300 { return 15 }
	if x >= 100 { return 10 }
	return 5
}

func bench(n int64) int64 {
	var s int64
	for i := int64(0); i < n; i++ {
		s += classify(i)
	}
	return s
}

func main() {
	n := int64(1000)
	if len(os.Args) > 1 {
		if v, err := strconv.ParseInt(os.Args[1], 10, 64); err == nil {
			n = v
		}
	}
	for i := 0; i < 1000; i++ {
		bench(n)
	}
	iters := int64(10000)
	start := time.Now()
	var r int64
	for i := int64(0); i < iters; i++ {
		r = bench(n)
	}
	elapsed := time.Since(start).Nanoseconds()
	fmt.Printf("result:     %d\n", r)
	fmt.Printf("iterations: %d\n", iters)
	fmt.Printf("total:      %.2fms\n", float64(elapsed)/1e6)
	fmt.Printf("per call:   %dns\n", elapsed/iters)
}

File: research/explorations/bench-realistic/guards.lua
local function classify(x)
    if x >= 900 then return 30 end
    if x >= 700 then return 25 end
    if x >= 500 then return 20 end
    if x >= 300 then return 15 end
    if x >= 100 then return 10 end
    return 5
end

local function bench(n)
    local s = 0
    for i = 0, n - 1 do
        s = s + classify(i)
    end
    return s
end

local N = tonumber(arg[1]) or 1000
-- warmup
for i = 1, 1000 do bench(N) end
local iters = 10000
local clock = os.clock
local start = clock()
local r
for i = 1, iters do r = bench(N) end
local elapsed = (clock() - start) * 1e9
io.write(string.format("result:     %d\n", r))
io.write(string.format("iterations: %d\n", iters))
io.write(string.format("total:      %.2fms\n", elapsed / 1e6))
io.write(string.format("per call:   %.0fns\n", elapsed / iters))
